crates/sats/src/slim_slice.rs

```diff
@@ -0,0 +1,1060 @@
+    }
+
+    /// Returns a limited shared slice to this boxed slice.
+    #[allow(clippy::needless_lifetimes)]
+    pub fn shared_ref<'a>(&'a self) -> &'a SlimSlice<'a, T> {
+        // SAFETY: The reference lives as long as `self`.
+        // By virtue of `repr(transparent)` we're also allowed these reference casts.
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Returns a limited mutable slice to this boxed slice.
+    #[allow(clippy::needless_lifetimes)]
+    pub fn exclusive_ref<'a>(&'a mut self) -> &'a mut SlimSliceMut<'a, T> {
+        // SAFETY: The reference lives as long as `self`
+        // and we have exclusive access to the heap data thanks to `&'a mut self`.
+        // By virtue of `repr(transparent)` we're also allowed these reference casts.
+        unsafe { mem::transmute(self) }
+    }
+}
+
+impl<T> Deref for SlimSliceBox<T> {
+    type Target = [T];
+
+    #[allow(clippy::needless_lifetimes)]
+    fn deref<'a>(&'a self) -> &'a Self::Target {
+        // SAFETY: `ptr` and `len` are derived from a live `Box<[T]>` valid for `'self`
+        // so we satisfy all safety requirements for `from_raw_parts`.
+        unsafe { self.raw.deref() }
+    }
+}
+
+impl<T> DerefMut for SlimSliceBox<T> {
+    #[allow(clippy::needless_lifetimes)]
+    fn deref_mut<'a>(&'a mut self) -> &'a mut Self::Target {
+        // SAFETY: `ptr` and `len` are derived from a live `Box<[T]>`
+        // so we satisfy all safety requirements for `from_raw_parts`.
+        unsafe { self.raw.deref_mut() }
+    }
+}
+
+impl<T> TryFrom<Box<[T]>> for SlimSliceBox<T> {
+    type Error = LenTooLong<Box<[T]>>;
+
+    fn try_from(boxed: Box<[T]>) -> Result<Self, Self::Error> {
+        ensure_len_fits!(boxed);
+        // SAFETY: Checked above that `len <= u32::MAX`.
+        Ok(unsafe { Self::from_boxed_unchecked(boxed) })
+    }
+}
+
+impl<T> TryFrom<Vec<T>> for SlimSliceBox<T> {
+    type Error = LenTooLong<Vec<T>>;
+
+    fn try_from(vec: Vec<T>) -> Result<Self, Self::Error> {
+        ensure_len_fits!(vec);
+        // SAFETY: Checked above that `len <= u32::MAX`.
+        Ok(unsafe { Self::from_boxed_unchecked(vec.into_boxed_slice()) })
+    }
+}
+
+impl<T, const N: usize> From<[T; N]> for SlimSliceBox<T> {
+    fn from(arr: [T; N]) -> Self {
+        // Make sure `N <= u32::MAX`.
+        struct AssertU32<const N: usize>;
+        impl<const N: usize> AssertU32<N> {
+            const OK: () = assert!(N <= u32::MAX as usize);
+        }
+        #[allow(clippy::let_unit_value)]
+        let () = AssertU32::<N>::OK;
+
+        // SAFETY: We verified statically by `AssertU32<N>` above that `N` fits in u32.
+        unsafe { Self::from_boxed_unchecked(into_box(arr)) }
+    }
+}
+
+impl<T> From<SlimSliceBox<T>> for Box<[T]> {
+    fn from(slice: SlimSliceBox<T>) -> Self {
+        let slice = ManuallyDrop::new(slice);
+        let (ptr, len) = slice.raw.split();
+        // SAFETY: `ptr` and `len` originally came from a `Box<[T]>`.
```

Comment:
I don't understand this SAFETY comment. I would have thought that callers are required to only ever create SlimSliceBoxed which respect the invariants, and therefore ptr is required to be a valid pointer to `len` many things.